### PR TITLE
Cleanup executor arg holder

### PR DESCRIPTION
**Pull request by zasdfgbnm**
I am currently facing two chicken-egg issue when trying to add fusion input support for allocation domain:
1. When binding the inputs, I need to call `bindInputForExprEvaluation`, where there are checks on the consistency about strides. However, without binding inputs, I will not be able to get an `ExpressionEvaluator` that compute the strides.
2. I need to get a `KernelArgumentHolder` object in order to bind inputs, however, `KernelArgumentHolder` needs information about strides during construction, which can not be evaluated without binding inputs.

To solve these chicken-egg problems, I need to do a refactor on executor logic. This is the first PR of this refactor.

In this PR, I refactored `TensorArgAbstract`, `TensorArg`, `TensorArgCodegen`, `CpuScalarTensorCodegen`:
- Data type templates in these classes are removed. What matters is the size of data type, not the data type itself. Therefore, the type dispatch logic is simplified.
- Now `TensorArgAbstract` has a `tensor_` but not `instance_`, and it is no longer purely abstract. When calling `getTensorArg` without index type, you will create an object of `TensorArgAbstract`, and when with index type, you will create an instance of `TensorArg`. I believe this is a cleaner solution than the `index_type_resolved_` approach. In the future, `TensorArgAbstract` will be used to hold tensor args whose strides and index types are not yet know, and `TensorArg` will be used to hold tensor args whose strides and index types are known.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Review by csarofeen — APPROVED** [review]
I don't understand why we need an ExprEval to determine stride information.

However, I don't see any issues with this refactor.

File: csrc/executor_kernel_arg.h [hunk 1]
```diff
@@ -285,35 +233,81 @@ struct TensorArg : public TensorArgAbstract {
...
+    auto rank = getRank();
+    ss << "tensor dtype: " << getDataType() << " sizes: (";
+    for (auto i = 0; i < rank; i++) {
+      ss << getSize(i) << ", ";
+    }
+    ss << ") pointer: " << getPointer();
+    return ss.str();
   }
 
   std::unique_ptr<ArgAbstract> copy_unique_ptr() const override {
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.h]: renamed

**Comment by zasdfgbnm** [comment]
> I don't understand why we need an ExprEval to determine stride information.

It is needed to support generic allocation domain. For example, if the allocation domain has a split of an ID in rFactor domain, we need to know the size of split.

**Comment by csarofeen** [comment]
Okay, yeah! duh! Thanks.

**Pull request merged**